pkg/store/consul/pcstore/consul_store_test.go

```diff
@@ -1015,6 +1020,126 @@ func TestWatchAndSync(t *testing.T) {
+	quit := make(chan struct{})
+	defer close(quit)
+	syncSignal := make(chan struct{})
+	defer close(syncSignal)
+	deleteSignal := make(chan struct{})
+	defer close(deleteSignal)
+
+	example := examplePodCluster()
+	pc1, err := store.Create(
+		example.PodID,
+		example.AvailabilityZone,
+		"name1",
+		example.PodSelector,
+		example.Annotations,
+		consultest.NewSession(),
+	)
+	if err != nil {
+		t.Fatalf("Couldn't create test pod cluster: %s", err)
+	}
+
+	// Include two pod clusters as initial, but we only created one. we
+	// should see a delete call for the second
+	syncer := &RecordingSyncer{
+		InitialClusters: []fields.ID{pc1.ID, "some_other_ID"},
+		SyncSignal:      syncSignal,
+		DeleteSignal:    deleteSignal,
+	}
+
+	go func() {
+		err := store.WatchAndSync(syncer, quit)
+		if err != nil {
+			t.Fatalf("Couldn't start WatchAndSync(): %s", err)
+		}
+	}()
+
+	// We want a sync for pc
+	expectedSyncCount := 1
+	actualSyncCount := 0
+	for actualSyncCount < expectedSyncCount {
+		select {
+		case <-time.After(1 * time.Second):
+			t.Fatalf("Timed out waiting for sync to happen, there were %d syncs: %s", actualSyncCount, syncer.SyncClusterCalls)
+		case <-syncSignal:
+			actualSyncCount++
+		}
+	}
+
+	found := false
+	for _, id := range syncer.SyncClusterCalls {
+		if id == pc1.ID {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		t.Fatalf("Expected sync to be called for %s but it wasn't. Called for %s", pc1.ID, syncer.SyncClusterCalls)
+	}
+
+	// We want a delete for "some_other_ID"
+	expectedDeleteCount := 1
+	actualDeleteCount := 0
+	for actualDeleteCount < expectedDeleteCount {
+		select {
+		case <-time.After(1 * time.Second):
+			t.Fatalf("Timed out waiting for delete to happen, there were %d delete: %s", actualDeleteCount, syncer.DeleteClusterCalls)
+		case <-deleteSignal:
+			actualDeleteCount++
+		}
+	}
+
+	found = false
+	for _, id := range syncer.DeleteClusterCalls {
+		if id == "some_other_ID" {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		t.Fatalf("Expected sync to be called for %s but it wasn't. Called for %s", "some_other_ID", syncer.DeleteClusterCalls)
+	}
+}
+
+// Tests that we don't delete any clusters when we find none as a failsafe
+func TestWatchAndSyncFailsafe(t *testing.T) {
+	store := consulStoreWithFakeKV()
+	quit := make(chan struct{})
+	defer close(quit)
+	syncSignal := make(chan struct{})
+	defer close(syncSignal)
+	deleteSignal := make(chan struct{})
+	defer close(deleteSignal)
+
+	// Include two initial clusters, but don't actually put any in the
+	// store. We shouldn't see any delete calls because of the failsafe
+	// protection against empty pod cluster set
+	syncer := &RecordingSyncer{
+		InitialClusters: []fields.ID{"some_id", "some_other_ID"},
+		DeleteSignal:    deleteSignal,
+	}
+
+	go func() {
+		err := store.WatchAndSync(syncer, quit)
+		if err != nil {
+			t.Fatalf("Couldn't start WatchAndSync(): %s", err)
+		}
+	}()
+
+	select {
+	case <-time.After(1 * time.Second):
```

Comment:
writing a test that must take a second to pass makes me sad but short of passing an error on a channel within the failsafe `if` to the syncing routine i'm not sure how to avoid it